packages/next/src/build/templates/pages.ts

```diff
@@ -52,10 +66,315 @@ export const routeModule = new PagesRouteModule({
+            err: getRequestMeta(req, 'invokeError'),
+            dev: routeModule.isDev,
+
+            // needed for experimental.optimizeCss feature
+            distDir: `${routeModule.projectDir}/${routeModule.distDir}`,
+
+            ampSkipValidation: nextConfig.experimental.amp?.skipValidation,
+            ampValidator: getRequestMeta(req, 'ampValidator'),
+          },
+        })
+        .finally(() => {
+          if (!span) return
+
+          span.setAttributes({
+            'http.status_code': res.statusCode,
+            'next.rsc': false,
+          })
+
+          const rootSpanAttributes = tracer.getRootSpanAttributes()
+          // We were unable to get attributes, probably OTEL is not enabled
+          if (!rootSpanAttributes) {
+            return
+          }
+
+          if (
+            rootSpanAttributes.get('next.span_type') !==
+            BaseServerSpan.handleRequest
+          ) {
+            console.warn(
+              `Unexpected root span type '${rootSpanAttributes.get(
+                'next.span_type'
+              )}'. Please report this Next.js issue https://github.com/vercel/next.js`
+            )
+            return
+          }
+
+          const route = rootSpanAttributes.get('next.route')
+          if (route) {
+            const name = `${method} ${route}`
+
+            span.setAttributes({
+              'next.route': route,
+              'http.route': route,
+              'next.span_name': name,
+            })
+            span.updateName(name)
+          } else {
+            span.updateName(`${method} ${req.url}`)
+          }
+        })
+
+    // TODO: activeSpan code path is for when wrapped by
+    // next-server can be removed when this is no longer used
+    if (activeSpan) {
+      return await invokeRouteModule(activeSpan)
+    } else {
+      return await tracer.withPropagatedContext(req.headers, () =>
+        tracer.trace(
+          BaseServerSpan.handleRequest,
+          {
+            spanName: `${method} ${req.url}`,
+            kind: SpanKind.SERVER,
+            attributes: {
+              'http.method': method,
+              'http.target': req.url,
+            },
+          },
+          invokeRouteModule
+        )
+      )
+    }
+  } catch (err) {
+    await onError(
+      req,
+      err,
+      {
+        path: req.url || '/',
+        headers: req.headers,
+        method: req.method || 'GET',
+      },
+      {
+        routerKind: 'Pages Router',
+        routePath: srcPage,
+        routeType: 'render',
+        revalidateReason: getRevalidateReason({
+          isRevalidate: hasStaticProps,
+          isOnDemandRevalidate,
+        }),
+      }
+    )
+
+    // rethrow so that we can handle serving error page
+    throw err
+  } finally {
+    // We don't allow any waitUntil work in pages API routes currently
+    // so if callback is present return with resolved promise since no
+    // pending work
+    ctx.waitUntil?.(Promise.resolve())
```

Comment:
Are we sure about this? Why do we need to wait for a promise to resolve here?

Reply:
Why not don't call it then if that's the case? I don't understand the blank promise being waited upon, seems like a bad smell.

Reply:
Not sure I follow. How would users cleanly handle the `waitUntil` callback passed in potentially never being called, they'd have to do something like waiting some ticks or something to be sure it's not just delayed which seems unreliable versus just calling with a resolved `Promise` when no pending work is in flight.